environs/manual/winrmprovisioner/winrmprovisioner.go

```diff
@@ -0,0 +1,581 @@
+	// we will use this in order to append into a .ps1 file.
+	var buf [1024]byte
+
+	// if the file dosen't exist ,create it
+	// if the file exists just clear/reset it
+	script, err = shell.NewPSEncodedCommand(initChunk)
+	if err != nil {
+		return err
+	}
+	if err = cli.Run(script, stdin, stderr); err != nil {
+		return errors.Trace(err)
+	}
+
+	// sequential read.
+	for input.Len() != 0 {
+		n, err := input.Read(buf[:])
+		if err != nil && err != io.EOF {
+			return errors.Trace(err)
+		}
+		script, err = shell.NewPSEncodedCommand(
+			fmt.Sprintf(saveChunk, string(buf[:n])),
+		)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if err = cli.Run(script, stdin, stderr); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	// after the sendAndSave script is sucessfully done
+	// we must execute the newly writed script
+	script, err = shell.NewPSEncodedCommand(runCmdProv)
+	if err != nil {
+		return err
+	}
+	logger.Debugf("Runing the provisioningScript")
+	var outerr bytes.Buffer
+	if err = cli.Run(script, stdin, &outerr); err != nil {
+		return errors.Trace(err)
+	}
+
+	if outerr.Len() != 0 {
+		return fmt.Errorf("%v ", strings.TrimSpace(outerr.String()))
+	}
+
+	return err
+}
+
+// initChunk creates or clears the file that the userdata will be appendend.
+const initChunk = `
+$provisionPath= [io.path]::Combine($ENV:APPDATA, 'Juju', 'provision.ps1')
+if (-Not (Test-Path $provisionPath)) {
+	New-Item $provisionPath -Type file
+} else {
+	Clear-Content $provisionPath
+}
+
+`
+
+// saveChunk powershell script that will append into the userdata file created after the
+// initChunk executed.
+// this will be called multiple times in a sequential order.
+const saveChunk = `
+$chunk= @"
+%s
+"@
+$provisionPath= [io.path]::Combine($ENV:APPDATA, 'Juju', 'provision.ps1')
+$stream = New-Object System.IO.StreamWriter -ArgumentList ([IO.File]::Open($provisionPath, "Append"))
+$stream.Write($chunk)
+$stream.close()
+`
+
+// runCmdProv powrshell script that decodes and executes the newly created userdata file
+// after the process of writing the sequantial script is done above
+const runCmdProv = `
+$provisionPath= [io.path]::Combine($ENV:APPDATA, 'Juju', 'provision.ps1')
+$script = [IO.File]::ReadAllText($provisionPath)
+$x = [System.Text.Encoding]::ASCII.GetString([System.Convert]::FromBase64String($script))
+Set-Content C:\udata.ps1 $x
+powershell.exe -ExecutionPolicy RemoteSigned -NonInteractive -File C:\udata.ps1
+`
+
+// ProvisioningScript generates a powershell script that can be
+// executed on a remote host to carry out the cloud-init
+// configuration.
+func ProvisioningScript(icfg *instancecfg.InstanceConfig) (string, error) {
+	cloudcfg, err := cloudinit.New(icfg.Series)
+	if err != nil {
+		return "", errors.Annotate(err, "error generating cloud-config")
+	}
+
+	udata, err := cloudconfig.NewUserdataConfig(icfg, cloudcfg)
+	if err != nil {
+		return "", errors.Annotate(err, "error generating cloud-config")
```

Comment:
ideally you'd make these error messages distinct so it's more obvious which one is causing the problem.